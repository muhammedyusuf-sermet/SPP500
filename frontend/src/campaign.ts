export interface ICampaignState {
	Id?: string,
	Name: string,
	Description: string;
	// Monsters: Monster.IMonsterState[];
}
