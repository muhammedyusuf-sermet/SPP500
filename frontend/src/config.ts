export const API_URL = "https://http://3.18.65.138:3000";