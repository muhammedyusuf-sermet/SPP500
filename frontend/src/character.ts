import { ICampaignData } from "./campaign";

export enum CharacterRace {
	Dragonborn = "Dragonborn",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Gnome = "Gnome",
	HalfElf = "HalfElf",
	Halfling = "Halfling",
	HalfOrc = "HalfOrc",
	Human = "Human",
	Tiefling = "Tiefling"
}

export enum CharacterClass {
	Barbarian = "Barbarian",
	Bard = "Bard",
	Cleric = "Cleric",
	Druid = "Druid",
	Fighter = "Fighter",
	Monk = "Monk",
	Paladin = "Paladin",
	Ranger = "Ranger",
	Rogue = "Rogue",
	Sorcerer = "Sorcerer",
	Warlock = "Warlock",
	Wizard = "Wizard"
}

export interface ICharacterData {
	Id?: number;
	Name: string;
	Level?: number;
	Race?: CharacterRace;
	Class?: CharacterClass;
	MaxHealth?: number;
	ArmorClass?: number;
	Notes?: string;

	Campaigns?: ICampaignData[];
}
