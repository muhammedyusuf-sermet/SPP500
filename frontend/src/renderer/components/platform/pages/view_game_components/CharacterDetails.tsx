import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import { CharacterRace, CharacterClass, ICharacterData } from '../../../../../character';
//import { ICampaignState } from '../../../../../campaign';
import { EnumDropdown } from '../../../helpers/EnumDropdown';
import { stateWithoutErrors } from '../../../../../utils/StateSelection';

const races = Object.values(CharacterRace);
const classes = Object.values(CharacterClass);

export interface IEncounterState {
	Id?: number
}

export interface ICharacterDetailsProps {
	disabled?: boolean,
	// validation
	PayloadSchema: JoiObject,
	ValidationOptions: ValidationOptions,
	// initial values
	initial: {
		Name?: string;
		Level?: number;
		Race?: CharacterRace;
		Class?: CharacterClass;
		MaxHealth?: number;
		ArmorClass?: number;
		Notes?: string;
		//Campaigns?: ICampaignState[];
	}
}

export interface ICharacterDetailsState {
	// data
	Name?: string;
	Level?: number;
	Race?: CharacterRace;
	Class?: CharacterClass;
	MaxHealth?: number;
	ArmorClass?: number;
	Notes?: string;
	//Campaigns?: string;
	// errors
	NameError?: string;
	LevelError?: string;
	RaceError?: string;
	ClassError?: string;
	MaxHealthError?: string;
	ArmorClassError?: string;
	NotesError?: string;
	//CampaignsError?: string;
}

export class CharacterDetails extends React.Component<ICharacterDetailsProps, ICharacterDetailsState> {
	constructor(props: ICharacterDetailsProps) {
		super(props);
		this.state = {
			...props.initial,
			//Campaigns: props.initial.Campaigns ? props.initial.Campaigns.map((value)=>(value.Id)).join(',') : ''
		};
	}

	componentWillReceiveProps(nextProps: ICharacterDetailsProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false) {
			this.setState({
				...nextProps.initial,
				//Campaigns: nextProps.initial.Campaigns ? nextProps.initial.Campaigns.map((value)=>(value.Id)).join(',') : ''
			});
		} else if (this.props.disabled != nextProps.disabled) {
			this.forceUpdate();
		}
	}

	handleStringChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = /^\s*$/.test(event.target.value) ? undefined : event.target.value;
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
		});
	}

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			value,
			Joi.reach(this.props.PayloadSchema, [name]),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
	}

	handleCharacterEnumChange = (name: string, newEnum?: string) => {
		if (newEnum) {
			Joi.validate(
				newEnum,
				Joi.reach(this.props.PayloadSchema, [name]),
				this.props.ValidationOptions,
				(errors: ValidationError) => {
					if(errors) {
						this.setState({
							[name+'Error']: errors.details[0].message
						});
					} else {
						this.setState({
							[name]: newEnum,
							[name+'Error']: undefined
						});
					}
				});
		} else {
			this.setState({
				[name]: undefined,
				[name+'Error']: undefined
			});
		}
	}

	public GetData = (): ICharacterData => {
		return stateWithoutErrors(this.state);
	}

	render() {
		return (
			<Grid container spacing={8} >
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Name">Name</InputLabel>
						<Input
							id='Name'
							type='text'
							value={this.state.Name || ''}
							name='Name'
							onChange={this.handleStringChange}
							aria-describedby="Name-helper-text" />
						<FormHelperText error id="Name-helper-text">{this.state.NameError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={6} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor='Race' >Race</InputLabel>
						<EnumDropdown
							disabled={this.props.disabled}
							selected={this.state.Race}
							name='Race'
							options={races}
							onChange={this.handleCharacterEnumChange} />
						<FormHelperText error id='Race-helper-text' >{this.state.RaceError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={6} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor='Class' >Class</InputLabel>
						<EnumDropdown
							disabled={this.props.disabled}
							selected={this.state.Class}
							name='Class'
							options={classes}
							onChange={this.handleCharacterEnumChange} />
						<FormHelperText error id='Class-helper-text' >{this.state.ClassError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Level">Level</InputLabel>
						<Input
							id='Level'
							type='number'
							value={this.state.Level != undefined ? this.state.Level : ''}
							name='Level'
							onChange={this.handleNumberChange}
							aria-describedby="Level-helper-text" />
						<FormHelperText error id="Level-helper-text">{this.state.LevelError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={6}>
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="ArmorClass">Armor Class</InputLabel>
						<Input
							id='ArmorClass'
							type='number'
							value={this.state.ArmorClass != undefined ? this.state.ArmorClass : ''}
							name='ArmorClass'
							onChange={this.handleNumberChange}
							aria-describedby="ArmorClass-helper-text" />
						<FormHelperText error id="ArmorClass-helper-text">{this.state.ArmorClassError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={6}>
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="MaxHealth">Max Hit Points</InputLabel>
						<Input
							id='MaxHealth'
							type='number'
							value={this.state.MaxHealth != undefined ? this.state.MaxHealth : ''}
							name='MaxHealth'
							onChange={this.handleNumberChange}
							aria-describedby="MaxHealth-helper-text" />
						<FormHelperText error id="MaxHealth-helper-text">{this.state.MaxHealthError}</FormHelperText>
					</FormControl>
				</Grid>
				<Grid item xs={12} >
					<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
						<InputLabel htmlFor="Notes">Notes</InputLabel>
						<Input
							id='Notes'
							type='text'
							multiline
							rows={10}
							rowsMax={10}
							value={this.state.Notes || ''}
							name='Notes'
							onChange={this.handleStringChange}
							aria-describedby="Notes-helper-text" />
						<FormHelperText error id="Notes-helper-text">{this.state.NotesError}</FormHelperText>
					</FormControl>
				</Grid>
			</Grid>
		);
	}
}