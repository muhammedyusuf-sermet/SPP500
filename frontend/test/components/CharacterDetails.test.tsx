import * as React from 'react';
import { mount, ReactWrapper, shallow } from 'enzyme';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

import { ICharacterDetailsProps, ICharacterDetailsState, CharacterDetails } from '../../src/renderer/components/platform/pages/view_game_components/CharacterDetails';
import { CharacterRace, CharacterClass } from '../../src/character';

const payloadSchema = Joi.object({
	Id: Joi.number().greater(0).allow(0),
	Name: Joi.string().required().max(50).label("Name"),
	Level: Joi.number().integer().greater(0).label("Level"),
	Race: Joi.string().valid(Joi.ref('$RaceOptions')).label('Race'),
	Class: Joi.string().valid(Joi.ref('$ClassOptions')).label('Class'),
	MaxHealth: Joi.number().integer().greater(0).label('MaxHealth'),
	ArmorClass: Joi.number().integer().greater(0).label('ArmorClass'),
	Notes: Joi.string().max(1000).label("Notes"),
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {
		RaceOptions: Object.keys(CharacterRace),
		ClassOptions: Object.keys(CharacterClass)
	}
};

////// Happy Path //////
describe('Character Details', () => {

	let monsterEnumConfigurationInstance: ReactWrapper<ICharacterDetailsProps, ICharacterDetailsState, CharacterDetails>;

	describe('Provided Character Details values', () => {

		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<CharacterDetails, ICharacterDetailsProps, ICharacterDetailsState>(
					<CharacterDetails
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							Name: 'Hello',
							Level: 3,
							Race: CharacterRace.Dragonborn,
							Class: CharacterClass.Cleric,
							MaxHealth: 35,
							ArmorClass: 17,
							Notes: 'Very ugly and ignorant'
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('should have the correct Name', () => {
			expect(monsterEnumConfigurationInstance.find('input#Name').props().value).toEqual('Hello');
		});

		it('should have the correct Level', () => {
			expect(monsterEnumConfigurationInstance.find('input#Level').props().value).toEqual(3);
		});

		it('should have the selected Race', () => {
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Dragonborn');
		});

		it('should have the selected Class', () => {
			expect(monsterEnumConfigurationInstance.find('select#Class').props().value).toEqual('Cleric');
		});

		it('should have the correct MaxHealth', () => {
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().value).toEqual(35);
		});

		it('should have the correct ArmorClass', () => {
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().value).toEqual(17);
		});

		it('should have the correct Notes', () => {
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().value).toEqual('Very ugly and ignorant');
		});
	});

	describe('componentWillReceiveProps()', () => {
		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<CharacterDetails, ICharacterDetailsProps, ICharacterDetailsState>(
					<CharacterDetails
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							Name: 'Hello',
							Level: 3,
							Race: CharacterRace.Dragonborn,
							Class: CharacterClass.Cleric,
							MaxHealth: 35,
							ArmorClass: 17,
							Notes: 'Very ugly and ignorant'
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('renders new props', () => {
			monsterEnumConfigurationInstance.setProps({
				initial: {
					Name: 'NewName',
					Level: 5,
					Race: CharacterRace.Elf,
					Class: CharacterClass.Sorcerer,
					MaxHealth: 50,
					ArmorClass: 12,
					Notes: 'Very cool and insane'
				}
			})
			expect(monsterEnumConfigurationInstance.find('input#Name').props().value).toEqual('NewName');
			expect(monsterEnumConfigurationInstance.find('input#Level').props().value).toEqual(5);
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Elf');
			expect(monsterEnumConfigurationInstance.find('select#Class').props().value).toEqual('Sorcerer');
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().value).toEqual(50);
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().value).toEqual(12);
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().value).toEqual('Very cool and insane');
		});

		it('renders without crashing with same props', () => {
			monsterEnumConfigurationInstance.setProps({
				initial: {
					Name: 'Hello',
					Level: 3,
					Race: CharacterRace.Dragonborn,
					Class: CharacterClass.Cleric,
					MaxHealth: 35,
					ArmorClass: 17,
					Notes: 'Very ugly and ignorant'
				}
			});
			expect(monsterEnumConfigurationInstance.find('input#Name').props().value).toEqual('Hello');
			expect(monsterEnumConfigurationInstance.find('input#Level').props().value).toEqual(3);
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Dragonborn');
			expect(monsterEnumConfigurationInstance.find('select#Class').props().value).toEqual('Cleric');
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().value).toEqual(35);
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().value).toEqual(17);
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().value).toEqual('Very ugly and ignorant');
		});

		it('disables input if disable prop changes', () => {
			expect(monsterEnumConfigurationInstance.find('input#Name').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('input#Level').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('select#Race').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('select#Class').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().disabled).toEqual(false)
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().disabled).toEqual(false)
			monsterEnumConfigurationInstance.setProps({
				disabled: true
			});
			expect(monsterEnumConfigurationInstance.find('input#Name').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('input#Level').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('select#Race').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('select#Class').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().disabled).toEqual(true)
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().disabled).toEqual(true)
		});

		it('Does not render twice when the props are the same.', () => {
			const prevState = monsterEnumConfigurationInstance.state();
			monsterEnumConfigurationInstance.instance().componentWillReceiveProps({
				disabled: false,
				PayloadSchema: payloadSchema,
				ValidationOptions: validateOptions,
				initial: {
					Name: 'Hello',
					Level: 3,
					Race: CharacterRace.Dragonborn,
					Class: CharacterClass.Cleric,
					MaxHealth: 35,
					ArmorClass: 17,
					Notes: 'Very ugly and ignorant'
				}
			});
			// check not just the value but the literal object
			// we want to make sure the component doesn't render
			// when the props don't change
			expect(monsterEnumConfigurationInstance.state()).toBe(prevState)
		});

		it('Rerender when only initial changes.', () => {
			monsterEnumConfigurationInstance.instance().componentWillReceiveProps({
				disabled: false,
				PayloadSchema: payloadSchema,
				ValidationOptions: validateOptions,
				initial: {
					Name: 'NewName',
					Level: 5,
					Race: CharacterRace.Elf,
					Class: CharacterClass.Sorcerer,
					MaxHealth: 50,
					ArmorClass: 12,
					Notes: 'Very cool and insane'
				}
			});
			monsterEnumConfigurationInstance.update()
			expect(monsterEnumConfigurationInstance.find('input#Name').props().value).toEqual('NewName');
			expect(monsterEnumConfigurationInstance.find('input#Level').props().value).toEqual(5);
			expect(monsterEnumConfigurationInstance.find('select#Race').props().value).toEqual('Elf');
			expect(monsterEnumConfigurationInstance.find('select#Class').props().value).toEqual('Sorcerer');
			expect(monsterEnumConfigurationInstance.find('input#MaxHealth').props().value).toEqual(50);
			expect(monsterEnumConfigurationInstance.find('input#ArmorClass').props().value).toEqual(12);
			expect(monsterEnumConfigurationInstance.find('textarea#Notes').props().value).toEqual('Very cool and insane');
		});
	});

	describe('Small snapshot', () => {
		it ('matches snapshot', () => {
			const shallowCharacterDetailsInstance =
				shallow<CharacterDetails, ICharacterDetailsProps, ICharacterDetailsState>(
					<CharacterDetails
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}}/>
				);
			expect(shallowCharacterDetailsInstance).toMatchSnapshot();
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterEnumConfigurationInstance =
				mount<CharacterDetails, ICharacterDetailsProps, ICharacterDetailsState>(
					<CharacterDetails
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterEnumConfigurationInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterEnumConfigurationInstance.find('input#Name').simulate('change', { target: { value: 'Hello' }})
			monsterEnumConfigurationInstance.find('input#Level').simulate('change', { target: { value: 4 }})
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: 'HalfElf' }})
			monsterEnumConfigurationInstance.find('select#Class').simulate('change', { target: { value: 'Wizard' }})
			monsterEnumConfigurationInstance.find('input#MaxHealth').simulate('change', { target: { value: 25 }})
			monsterEnumConfigurationInstance.find('input#ArmorClass').simulate('change', { target: { value: 7 }})
			monsterEnumConfigurationInstance.find('textarea#Notes').simulate('change', { target: { value: 'Gray' }})
			expect(monsterEnumConfigurationInstance.state()).toEqual({
				// data
				Name: 'Hello',
				Level: 4,
				Race: 'HalfElf',
				Class: 'Wizard',
				MaxHealth: 25,
				ArmorClass: 7,
				Notes: 'Gray',
				// errors
				NameError: undefined,
				LevelError: undefined,
				RaceError: undefined,
				ClassError: undefined,
				MaxHealthError: undefined,
				ArmorClassError: undefined,
				NotesError: undefined,
			});
		});

		it('should be able to get only data', () => {
			monsterEnumConfigurationInstance.find('input#Name').simulate('change', { target: { value: 'Hello' }})
			// level is not set
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: 'HalfElf' }})
			// class is not set
			monsterEnumConfigurationInstance.find('input#MaxHealth').simulate('change', { target: { value: 25 }})
			monsterEnumConfigurationInstance.find('input#ArmorClass').simulate('change', { target: { value: 7 }})
			monsterEnumConfigurationInstance.find('textarea#Notes').simulate('change', { target: { value: 'Gray' }})
			// expect only data that was set
			expect(monsterEnumConfigurationInstance.instance().GetData()).toEqual({
				// data
				Name: 'Hello',
				Race: 'HalfElf',
				MaxHealth: 25,
				ArmorClass: 7,
				Notes: 'Gray',
			});
		});
		// STATE CHANGE
		it('should change only Name when Name is changed', () => {
			monsterEnumConfigurationInstance.find('input#Name').simulate('change', { target: { value: 'Hello' }})
			expect(monsterEnumConfigurationInstance.state().Name).toEqual('Hello');
		})

		it('should change only Level when Level is changed', () => {
			monsterEnumConfigurationInstance.find('input#Level').simulate('change', { target: { value: 4 }})
			expect(monsterEnumConfigurationInstance.state().Level).toEqual(4);
		})

		it('should change only Race when Race is changed', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: 'HalfElf' }})
			expect(monsterEnumConfigurationInstance.state().Race).toEqual('HalfElf');
		})

		it('should change only Class when Class is changed', () => {
			monsterEnumConfigurationInstance.find('select#Class').simulate('change', { target: { value: 'Wizard' }})
			expect(monsterEnumConfigurationInstance.state().Class).toEqual('Wizard');
		})

		it('should change only MaxHealth when MaxHealth is changed', () => {
			monsterEnumConfigurationInstance.find('input#MaxHealth').simulate('change', { target: { value: 45 }})
			expect(monsterEnumConfigurationInstance.state().MaxHealth).toEqual(45);
		})

		it('should change only ArmorClass when ArmorClass is changed', () => {
			monsterEnumConfigurationInstance.find('input#ArmorClass').simulate('change', { target: { value: 6 }})
			expect(monsterEnumConfigurationInstance.state().ArmorClass).toEqual(6);
		})

		it('should change only Notes when Notes is changed', () => {
			monsterEnumConfigurationInstance.find('textarea#Notes').simulate('change', { target: { value: 'A long list of attributes' }})
			expect(monsterEnumConfigurationInstance.state().Notes).toEqual('A long list of attributes');
		})
		// DEFAULTS
		it('should reset Name when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('input#Name').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().Name).toBeUndefined()
		})

		it('should reset Level when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('input#Level').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().Level).toBeUndefined()
		})

		it('should reset Race when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().Race).toBeUndefined()
		})

		it('should reset Class when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('select#Class').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().Class).toBeUndefined()
		})

		it('should reset MaxHealth when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('input#MaxHealth').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().MaxHealth).toBeUndefined()
		})

		it('should reset ArmorClass when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('input#ArmorClass').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().ArmorClass).toBeUndefined()
		})

		it('should reset Notes when input is set to empty', () => {
			monsterEnumConfigurationInstance.find('textarea#Notes').simulate('change', { target: { value: '' }})
			expect(monsterEnumConfigurationInstance.state().Notes).toBeUndefined()
		})
		// VALIDATION
		const longerThan50='NameTooLong01234567890123456789012345678901234567890123456789'

		it('should validate input when Name is changed', () => {
			monsterEnumConfigurationInstance.find('input#Name').simulate('change', { target: { value: longerThan50 }})
			expect(monsterEnumConfigurationInstance.state().Name).toEqual(longerThan50)
			expect(monsterEnumConfigurationInstance.state().NameError).toEqual('"Name" length must be less than or equal to 50 characters long')
		})

		it('should validate input when Level is changed', () => {
			monsterEnumConfigurationInstance.find('input#Level').simulate('change', { target: { value: -1 }})
			expect(monsterEnumConfigurationInstance.state().Level).toEqual(-1)
			expect(monsterEnumConfigurationInstance.state().LevelError).toEqual('"Level" must be greater than 0')
		})

		it('should validate input when Race is changed', () => {
			monsterEnumConfigurationInstance.find('select#Race').simulate('change', { target: { value: 'InvalidRace' }})
			expect(monsterEnumConfigurationInstance.state().Race).toBeUndefined()
			expect(monsterEnumConfigurationInstance.state().RaceError).toEqual('"Race" must be one of [context:RaceOptions]')
		})

		it('should validate input when Class is changed', () => {
			monsterEnumConfigurationInstance.find('select#Class').simulate('change', { target: { value: 'InvalidClass' }})
			expect(monsterEnumConfigurationInstance.state().Class).toBeUndefined()
			expect(monsterEnumConfigurationInstance.state().ClassError).toEqual('"Class" must be one of [context:ClassOptions]')
		})

		it('should validate input when MaxHealth is changed', () => {
			monsterEnumConfigurationInstance.find('input#MaxHealth').simulate('change', { target: { value: -1 }})
			expect(monsterEnumConfigurationInstance.state().MaxHealth).toEqual(-1)
			expect(monsterEnumConfigurationInstance.state().MaxHealthError).toEqual('"MaxHealth" must be greater than 0')
		})

		it('should validate input when ArmorClass is changed', () => {
			monsterEnumConfigurationInstance.find('input#ArmorClass').simulate('change', { target: { value: -1 }})
			expect(monsterEnumConfigurationInstance.state().ArmorClass).toEqual(-1)
			expect(monsterEnumConfigurationInstance.state().ArmorClassError).toEqual('"ArmorClass" must be greater than 0')
		})

		const longerThan1000 = "NotesLongerThan1000\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789\
			0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789"

		it('should validate input when Notes is changed', () => {
			monsterEnumConfigurationInstance.find('textarea#Notes').simulate('change', { target: { value: longerThan1000 }})
			expect(monsterEnumConfigurationInstance.state().Notes).toEqual(longerThan1000)
			expect(monsterEnumConfigurationInstance.state().NotesError).toEqual('"Notes" length must be less than or equal to 1000 characters long')
		})
	});
});