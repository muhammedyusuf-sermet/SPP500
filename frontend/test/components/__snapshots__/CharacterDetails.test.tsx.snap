// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Details Small snapshot matches snapshot 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={8}
>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="Name"
      >
        Name
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        aria-describedby="Name-helper-text"
        id="Name"
        name="Name"
        onChange={[Function]}
        type="text"
        value=""
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="Name-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={6}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="Race"
      >
        Race
      </WithStyles(WithFormControlContext(InputLabel))>
      <EnumDropdown
        disabled={false}
        name="Race"
        onChange={[Function]}
        options={
          Array [
            "Dragonborn",
            "Dwarf",
            "Elf",
            "Gnome",
            "HalfElf",
            "Halfling",
            "HalfOrc",
            "Human",
            "Tiefling",
          ]
        }
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="Race-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={6}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="Class"
      >
        Class
      </WithStyles(WithFormControlContext(InputLabel))>
      <EnumDropdown
        disabled={false}
        name="Class"
        onChange={[Function]}
        options={
          Array [
            "Barbarian",
            "Bard",
            "Cleric",
            "Druid",
            "Fighter",
            "Monk",
            "Paladin",
            "Ranger",
            "Rogue",
            "Sorcerer",
            "Warlock",
            "Wizard",
          ]
        }
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="Class-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="Level"
      >
        Level
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        aria-describedby="Level-helper-text"
        id="Level"
        name="Level"
        onChange={[Function]}
        type="number"
        value=""
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="Level-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={6}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="ArmorClass"
      >
        Armor Class
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        aria-describedby="ArmorClass-helper-text"
        id="ArmorClass"
        name="ArmorClass"
        onChange={[Function]}
        type="number"
        value=""
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="ArmorClass-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={6}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="MaxHealth"
      >
        Max Hit Points
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        aria-describedby="MaxHealth-helper-text"
        id="MaxHealth"
        name="MaxHealth"
        onChange={[Function]}
        type="number"
        value=""
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="MaxHealth-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(FormControl)
      className="formControl"
      disabled={false}
      fullWidth={true}
    >
      <WithStyles(WithFormControlContext(InputLabel))
        htmlFor="Notes"
      >
        Notes
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        aria-describedby="Notes-helper-text"
        id="Notes"
        multiline={true}
        name="Notes"
        onChange={[Function]}
        rows={10}
        rowsMax={10}
        type="text"
        value=""
      />
      <WithStyles(WithFormControlContext(FormHelperText))
        error={true}
        id="Notes-helper-text"
      />
    </WithStyles(FormControl)>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
